xml/System/Object.xml

```diff
@@ -685,7 +685,7 @@ and the <xref:System.IDisposable> interface. The <xref:System.IDisposable.Dispos
   
  Otherwise, you might think that the mutable object is lost in the hash table. If you do choose to override <see cref="M:System.Object.GetHashCode" /> for a mutable reference type, your documentation should make it clear that users of your type should not modify object values while the object is stored in a hash table.  
   
- For value types, <see cref="M:System.ValueType.GetHashCode" /> provides a default hash code implementation that uses reflection. You should consider overriding it for better performance.  
```

Comment:
I would keep the existing description.

The fast path in the default implementation is still fundamentally using reflection and it is still significantly slower than overridden implementation. I would not complicate the explanation here with runtime implementation details. Also, these details vary between runtime versions.

Reply:
I agree with Jan.  Following the chain of guidelines:

> DO implement IEquatable<T> on value types.

> DO override GetHashCode whenever you override Equals.

Our recommendations are that essentially every value type should explicitly override GetHashCode.  For some transient types (like one discussed in API review today) where they're just a custom type representing an N-tuple response from a callback, it might be reasonable to decide that value comparison is irrelevant and skip those... but then the performance implications of reflection don't matter.

Reply:
> The fast path in the default implementation is still fundamentally using reflection and it is still significantly slower than overridden implementation.

Hmm, but it's not using reflection on every call to `GetHashCode`, just once as set-up that is then cached. It makes every effort to be as performant as possible:

* Check if it is a `null` reference.
* Wrap object reference in `OBJECTREF` struct on the stack. This struct is just one pointer in size and is a debugging aid. Probably doesn't actually result in actual code.
* Get a pointer to the object's method table. Pretty sure this is O(1).
* Get the current AppDomain. This is O(1).
* Look up the TypeID. This is O(lg(N)) where N is the number of loaded types.
* Create an initial hash code by multiplying the TypeID by a big prime and applying an offset.
* Read the cached `CanCompareBitsOrUseFastGetHashCode` value.
* If `canUseFastGetHashCodeHelper` is true, then `FastGetValueTypeHashCodeHelper` reads the structure's data as though it were an unmanaged `int` array.

The majority of these are reading a single pointer from memory.

The first time you call it, it needs to compute the `CanCompareBitsOrUseFastGetHsahCode` value to cache, but the common path does _not_ use reflection to enumerate the members of the type. The total complexity is a map lookup where the map size is the number of loaded types and walking through the structure's raw data.

It is disingenuous to tell the user that `GetHashCode` calls _will by definition_ always use reflection. The argument seems to be that the user should be told this so that they'll feel more of an incentive to follow guidelines and provide their own `GetHashCode` implementation, but this is literally lying about what the implementation is doing for this purpose.

I just made a brute-force benchmark of it with a `struct` with 256 `int` fields, and I found:

* The default `GetHashCode` implementation takes about 2.7 times longer than a one-liner that XORs all the fields in a hardcoded manner.
* Using reflection to enumerate the fields and obtain each field's value takes about 123 times longer than the default `GetHashCode` implementation, or 338 times longer than the single expression.

![image](https://user-images.githubusercontent.com/2349379/161421596-b47c842f-35f1-4d95-b8ea-dcdd7c4038e3.png)